File: apple/SurgeFrameworks/Shared/Decoders/Utilities/BitstreamWriter.m
// Copyright (c) 2017 Instil Software.
//
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.

#import "BitstreamWriter.h"

@interface BitstreamWriter ()
@property (nonatomic, assign) uint8_t *buffer;
@property (nonatomic, assign) uint32_t currentBitOffset;
@end

@implementation BitstreamWriter

- (instancetype)init {
    self = [super init];
    if (self) {
        self.buffer = calloc(1024, sizeof(uint8_t));
    }
    return self;
}

- (void)dealloc {
    free(self.buffer);
}

- (void)writeBit:(uint8_t)value {
    [self writeBits:value numberOfBits:1];
}

- (void)writeByte:(uint8_t)value {
    [self writeBits:value numberOfBits:8];
}

- (void)writeTwoBytes:(uint16_t)value {
    [self writeByte:(value >> 8)];
    [self writeByte:value];
}

- (void)writeThreeBytes:(uint32_t)value {
    [self writeByte:(value >> 16)];
    [self writeByte:(value >> 8)];
    [self writeByte:value];
}

- (void)writeFourBytes:(uint32_t)value {
    [self writeByte:(value >> 24)];
    [self writeByte:(value >> 16)];
    [self writeByte:(value >> 8)];
    [self writeByte:value];
}

- (void)writeBytes:(NSData*)data {
    uint8_t *bytes = (uint8_t *)[data bytes];
    for (int i = 0; i < [data length]; i++) {
        [self writeByte:bytes[i]];
    }
}

- (void)writeBits:(uint8_t)value numberOfBits:(uint8_t)numberOfBits {
    uint32_t byteOffset = self.currentBitOffset / 8;
    uint8_t bitOffset = self.currentBitOffset % 8;
    uint8_t maxBitsWritable = 8 - bitOffset;
    if (numberOfBits <= maxBitsWritable) {
        self.buffer[byteOffset] |= value << (maxBitsWritable - numberOfBits);
    } else {
        self.buffer[byteOffset] |= value >> bitOffset;
        self.buffer[byteOffset + 1] = value << (8 - bitOffset);
    }
    self.currentBitOffset += numberOfBits;
}

- (NSData*)data {
    return [NSData dataWithBytes:self.buffer length:self.currentBitOffset / 8];
}

@end

File: core/src/BitstreamWriter.cc
// Copyright (c) 2017 Instil Software.
//
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.

#include "BitstreamWriter.h"

SurgeUtil::BitstreamWriter::BitstreamWriter() {
    bitstream = new std::vector<uint8_t>(1);
}

SurgeUtil::BitstreamWriter::~BitstreamWriter() {
    delete bitstream;
}

void SurgeUtil::BitstreamWriter::WriteBit(uint8_t value) {
    WriteBits(value, 1);
}

void SurgeUtil::BitstreamWriter::WriteByte(uint8_t value) {
    WriteBits(value, 8);
}

void SurgeUtil::BitstreamWriter::WriteTwoBytes(uint16_t value) {
    WriteByte(value >> 8);
    WriteByte(value);
}

void SurgeUtil::BitstreamWriter::WriteThreeBytes(uint32_t value) {
    WriteByte(value >> 16);
    WriteByte(value >> 8);
    WriteByte(value);
}

void SurgeUtil::BitstreamWriter::WriteFourBytes(uint32_t value) {
    WriteByte(value >> 24);
    WriteByte(value >> 16);
    WriteByte(value >> 8);
    WriteByte(value);
}

void SurgeUtil::BitstreamWriter::WriteBytes(uint8_t *bytes, size_t size) {
    for (int i = 0; i < size; i++) {
        WriteByte(bytes[i]);
    }
}

void SurgeUtil::BitstreamWriter::WriteBits(uint8_t value, uint8_t numberOfBits) {
    uint32_t byteOffset = currentBitOffset / 8;
    uint8_t bitOffset = currentBitOffset % 8;
    uint8_t maxBitsWritable = 8 - bitOffset;
    if (numberOfBits <= maxBitsWritable) {
        bitstream->data()[byteOffset] |= value << (maxBitsWritable - numberOfBits);
    } else {
        bitstream->resize(bitstream->size() + 1);
        bitstream->data()[byteOffset] |= value >> bitOffset;
        bitstream->data()[byteOffset + 1] = value << (8 - bitOffset);
    }
    currentBitOffset += numberOfBits;
}

std::vector<uint8_t>* SurgeUtil::BitstreamWriter::Bitstream() {
    return bitstream;
}
